### PR TITLE
PLFM-6008 STS APIs

### PR DESCRIPTION
**Pull request by DwayneJengSage**
https://sagebionetworks.jira.com/browse/PLFM-6008

Also manually tested with my own S3 bucket with no base key.'

Update: I also manually tested this with a bucket owned by Bridge (ie, the Synapse Dev account can get temporary credentials for a bucket owned by another account, assuming the steps in https://docs.synapse.org/articles/custom_storage_location.html are followed).

### CODE REVIEW

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/sts/StsManagerImpl.java [hunk 1]
```diff
@@ -1,27 +1,160 @@
...
+					.build();
+
+	// The AWS IAM policy string for the actions the user is allowed to do.
+	private static final Map<StsPermission, String> PERMISSION_TO_POLICY_ACTIONS =
+			ImmutableMap.<StsPermission, String>builder()
+					.put(StsPermission.read_only, "\"s3:Get*\",\"s3:List*\"")
+					.put(StsPermission.read_write, "\"s3:*\"")
+					.build();
+
+	private static final String SESSION_POLICY_TEMPLATE = "{\n" +
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/sts/StsManagerImpl.java]: move to a classpath resource.
Maybe use velocity (not sure if that overkill) https://velocity.apache.org/engine/2.0/user-guide.html
and we will review an example of a policy with IT.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/sts/StsManagerImpl.java [hunk 2]
```diff
@@ -1,27 +1,160 @@
...
 
 @Component
 public class StsManagerImpl implements StsManager {
+	static final int DURATION_SECONDS = 12 * 60 * 60; // 12 hours
+
+	// The Synapse ACCESS_TYPE for the given STS Permission. Used to determine if a user has the proper ACLs.
+	private static final Map<StsPermission, ACCESS_TYPE> PERMISSION_TO_ACCESS_TYPE =
+			ImmutableMap.<StsPermission, ACCESS_TYPE>builder()
+					.put(StsPermission.read_only, ACCESS_TYPE.DOWNLOAD)
+					.put(StsPermission.read_write, ACCESS_TYPE.UPLOAD)
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/sts/StsManagerImpl.java]: upload and download

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/sts/StsManagerImpl.java [hunk 3]
```diff
@@ -1,27 +1,160 @@
...
+			bucketWithFolder = bucket;
+			folderWithTrailingSlash = "";
+		} else {
+			bucketWithFolder = bucket + "/" + baseKey;
+			folderWithTrailingSlash = baseKey + "/";
+		}
+
+		// Call STS.
+		String actions = PERMISSION_TO_POLICY_ACTIONS.get(permission);
+		String policy = SESSION_POLICY_TEMPLATE.replace("$bucketWithFolder", bucketWithFolder)
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/sts/StsManagerImpl.java]: consider velocity to solve this problem.

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/sts/StsManagerImplTest.java [hunk 4]
```diff
@@ -1,48 +1,287 @@
...
+		// Mock dependencies.
+		setupFolderWithProjectSetting(/*isSts*/ true, STS_STORAGE_LOCATION_ID);
+
+		ExternalS3StorageLocationSetting storageLocationSetting = new ExternalS3StorageLocationSetting();
+		storageLocationSetting.setBucket(BUCKET);
+		storageLocationSetting.setStsEnabled(true);
+		when(mockProjectSettingsManager.getStorageLocationSetting(STS_STORAGE_LOCATION_ID)).thenReturn(
+				storageLocationSetting);
+
+		when(mockAuthManager.canAccess(USER_INFO, PARENT_ENTITY_ID, ObjectType.ENTITY, ACCESS_TYPE.UPLOAD))
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/sts/StsManagerImplTest.java]: can download too

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/sts/StsManagerImplAutowiredTest.java [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+	}
+
+	@Test
+	public void externalS3() throws Exception {
+		// Setup - In our test folder, create an inner folder. STS will be configured on the inner folder. STS
+		// credentials will be able to access files in the inner folder, but not the outer folder.
+		String outerFolderPath = "integration-test/" + "StsManagerImplAutowiredTest-" + UUID.randomUUID().toString();
+		String innerFolderPath = outerFolderPath + "/storage-location-root";
+		uploadFileToS3(outerFolderPath, "inaccessible.txt", "Dummy file content");
+		uploadFileToS3(innerFolderPath, "owner.txt", username);
```
Comment (DwayneJengSage) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/sts/StsManagerImplAutowiredTest.java]: From code review discussion, we wanted to make sure we included the following test cases:

1. inaccessible.txt at the root of the S3 bucket
2. baseKey/owner.txt
3. baseKey/subfolder/file.txt

Should test both listing the folder and getting the file.

Also, we don't need a second level of folders for the baseKey.

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/sts/StsManagerImplAutowiredTest.java [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+				EXTERNAL_S3_BUCKET, outerFolderPath + "/inaccessible.txt"));
+		assertEquals(403, ex.getStatusCode());
+
+		// Validate can write to S3. This call will not throw.
+		readWriteTempClient.putObject(EXTERNAL_S3_BUCKET, innerFolderPath + "/" + filenameToWrite,
+				"lorem ipsum");
+	}
+
+	@Test
+	public void synapseStorage() throws Exception {
```
Comment (DwayneJengSage) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/sts/StsManagerImplAutowiredTest.java:223]: For Synapse storage, we wanted to test both a FileHandle in our storage location (accessible) and a FileHandle in a different storage location (inaccessible).

**Review by john-hill — APPROVED** [review]

**Pull request merged**